lib/redis_dedupe/set.rb

```diff
@@ -0,0 +1,103 @@
+# frozen_string_literal: true
+
+# :nodoc:
+module RedisDedupe
+  class << self
+    attr_accessor :client
+  end
+
+  # A mechanism to make sure that a block of code will only be called once for each specified identifier (or +member+)
+  # even if the calling process dies or restarts, as long the datastore is +Redis+-backed.
+  #
+  # @example Keep the set around for 7 days, letting Redis handle its own memory cleanup after that time
+  #   ```
+  #   dedupe = RedisDedupe::Set.new($redis, "send_payment_due_emails")
+  #   Account.all do |account|
+  #     dedupe.check(account.id) do
+  #       mail(to: account.billing_email, subject: "PAY US... NOW!!!")
+  #     end
+  #   end
+  #   ```
+  #
+  # @example If you want to be able to repeat the process at any time immediately following this method
+  #   ```
+  #   dedupe = RedisDedupe::Set.new($redis, "send_welcome_emails")
+  #   Account.all.pluck(:email) do |email|
+  #     dedupe.check(email) { mail(to: email, subject: "Hello!") }
+  #   end
+  #   dedupe.finish
+  #   ```
+  #
+  class Set
+    SEVEN_DAYS = 7 * 24 * 60 * 60
+
+    attr_reader :key, :expires_in
+
+    def initialize(redis, key, expires_in = SEVEN_DAYS)
+      @redis      = redis
+      @key        = key
+      @expires_in = expires_in
+    end
+
+    # Ensures that a block of code will only be run if the +member+ is not already contained in Redis.
+    # ie: the code block has not already run for the specified +member+.
+    #
+    # Note that if the given block raises an error, the +member+ will not remain in the +Set+ and may be tried again.
+    #
+    # @param [String, Integer] member identifiying value to make sure the given block only runs once
+    #
+    # @yield block to run for the specified +member+, which should only be run once for any particular member
+    #
+    # @return `nil` if the block was not run, otherwise the result of the yielded block
+    #
+    def check(member, &block)
+      raise ArgumentError, "passing a block is required" if block.nil?
+      return nil unless execute_block_for_member?(member)
+
+      begin
+        block.call
+      rescue StandardError => e
+        redis.srem(key, member)
```

Comment:
So this is the change I'm still wrapping my head around. I think the behavior looks right compared to what we have (in fact, what we have I'd say is probably a bug).

I think the biggest trail my brain is going down is, could this have a negative impact on any existing app flows? I'm thinking for instance, if we run a large execution of a collection of records having to do with mailing, and there's a failure partway through the batch that is repeatable, that execution job would never be able to finish the collection. Currently, that becomes a "skip" and the collection can finish (this has its own problems, what if we are skipping something _really_ important). As a side note, another mechanism that comes to mind is some type of dead-letter queue.

I'm thinking one thing worth it would be to run the change by some teams with big collection/batch processes that rely on dedupe. Manage and Messaging teams come to mind. I'm interested in getting their thoughts on the behavioral change as it could impact their domains. /cc @cjbuchmann @thunderd0m3 @aburka @cforcey

Reply:
This is a very interesting proposed change.  I agree that what we have is essentially a `try_once_and_skip_on_exception` sort of functionality from this class which might come as a surprise to some consumers of the gem.  Your version feels closer to `execute_successfully_once` functionality that allows for transient errors sending an email to be recovered from by clearing the key from the set.

I am guessing your nervousness -- and mine -- comes from the wide-open nature of the block being wrapped by this helper.  If it is a pipeline has several steps each with side effects, then an exception on a later step might trigger a job retry duplicating side effects up to that point.  This feels a little far-fetched if what is being wrapped is a single invocation of a mailer. but some of our usages look a bit more complicated.

The API for this class is already a little odd -- `RedisDedupe::Set#check` does not roll off the tongue as a well-understood guarantee like `run_only_once` or  `run_successfully_only_once` might.  It might be total overkill, but you could offer an additional method in the current naming scheme like `check_finished` that wraps the `check` method in your rescue block.  

That would allow consumers of this library to decide whether it is more important to never run this trice -- exceptionally or successfully -- or to tolerate a few retries because it is important that at least one (but perhaps more) succeed.  The downside of that, of course, is that the new better behavior does not magically get applied in most of the cases where you would want it to.  You could also make your method `check` and have it wrap `check_if_duplicate_or_try_once` or some such method.  If someone really cared about the old behavior they could use that method (horribly named).

My last thought is that one RedisDedup is unlikely to handle all the different use cases for `run_once` functionality -- marketing messages already use email_deliveries to try to sort out what was sent in a first attempt and what should be retried.  Our rewrites of the transactional email system will most likely involve a more nuanced record like email_delivery for those messages that track the state of a delivery attempt and use that state to prevent the dreaded multiple emails problem.  Your PR inspired this small problem statement along those lines -- feedback is welcome!  https://kajabi.atlassian.net/wiki/spaces/MAR/pages/1234141598/Duplicate+Email+Prevention+in+Transactional+Email